### DISCUSSION TITLE
error(gtk_surface): surface failed to realize: error.PermissionDenied

### DISCUSSION THREAD

**Discussion by frap129** [ghostty-org/ghostty#3267] [Q&A] [answered]
Hello! I would like to test out ghostty, but am having issues getting it to start. I am running ArchLinux w/ Hyprland, and installed ghostty from the arch repos using yay. I found issue #1977 which describes the same issue I am having, without any fix or root cause.  Whenever I try to run ghostty, it logs the following and does not open a window:
```
❯ ghostty
info: ghostty version=1.0.0-dev+0000000
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=10.1.0
info: dependency fontconfig=21500
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=C.UTF-8
info(gtk): GTK version build=4.16.7 runtime=4.16.7
info(sentry): sentry envelope does not contain crash, discarding
info: optional config file not found, not loading path=/home/joe/.config/ghostty/config
info(config): default shell source=env value=/usr/bin/zsh
info(gtk): libadwaita version build=1.6.2 runtime=1.6.2

EGL API version 1.5 found
 - Vendor: Mesa Project
 - Version: 1.5
 - Device: /dev/dri/renderD128, 226 128
 - Client APIs: OpenGL OpenGL_ES
 - Extensions:
	EGL_ANDROID_blob_cache
	EGL_ANDROID_native_fence_sync
	EGL_EXT_buffer_age
	EGL_EXT_config_select_group
	EGL_EXT_create_context_robustness
	EGL_EXT_image_dma_buf_import
	EGL_EXT_image_dma_buf_import_modifiers
	EGL_EXT_pixel_format_float
	EGL_EXT_present_opaque
	EGL_EXT_protected_content
	EGL_EXT_query_reset_notification_strategy
	EGL_EXT_surface_compression
	EGL_EXT_swap_buffers_with_damage
	EGL_IMG_context_priority
	EGL_KHR_cl_event2
	EGL_KHR_config_attribs
	EGL_KHR_context_flush_control
	EGL_KHR_create_context
	EGL_KHR_create_context_no_error
	EGL_KHR_fence_sync
	EGL_KHR_get_all_proc_addresses
	EGL_KHR_gl_colorspace
	EGL_KHR_gl_renderbuffer_image
	EGL_KHR_gl_texture_2D_image
	EGL_KHR_gl_texture_3D_image
	EGL_KHR_gl_texture_cubemap_image
	EGL_KHR_image_base
	EGL_KHR_no_config_context
	EGL_KHR_reusable_sync
	EGL_KHR_surfaceless_context
	EGL_KHR_swap_buffers_with_damage
	EGL_KHR_wait_sync
	EGL_MESA_configless_context
	EGL_MESA_drm_image
	EGL_MESA_gl_interop
	EGL_MESA_image_dma_buf_export
	EGL_MESA_query_driver
	EGL_MESA_x11_native_visual_id
	EGL_WL_bind_wayland_display
	EGL_WL_create_wayland_buffer_from_image
 - Selected fbconfig: R8G8B8A8, depth 0, stencil 0
          high depth: R16G16B16A16 float, depth 0, stencil 0
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
Using OpenGL backend EGL
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
Creating EGL context version 3.3 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
info(grid): loaded OpenGL 4.6
error(gtk_surface): surface failed to realize: error.PermissionDenied
Destroying EGL context
```

**Comment by 00-kat** [comment, marked as answer]
Are you using the hardened kernel?  If not, strace it, to find out what actually fails.

Edit: the following is **no longer relevant**, see the next comment and #5916. TL;DR: if you have a **tip build**, you don't need to do anything.

---

If you are, the hardened kernel disables `io_uring` which is needed by `libxev` (#1977 was actually my issue but I didn't know why at that time).  You *can* change it to `Epoll`, but I don't think there's currently a build flag to do that.  (Related: https://github.com/mitchellh/libxev/issues/131.)  I'm currently using a hacky `sed`:
```shell
sed -i 's/^const xev = @import("xev");$/const xev = @import("xev").Epoll;/' **/*.zig
```
(P.S.: If you use bash, don't forget to enable recursive globbing with `shopt -s globstar`.)

**Reply by frap129** [reply]
Not using the hardened kernel, but had io_uring disabled via sysctl. `sudo sysctl kernel.io_uring_disabled=0` solved it for me, thanks!

**Reply by ssergiienko** [reply]
I can confirm ghostty currently depend on uring enabled, but re-compiling with epoll using the recipe [surprisingly](https://github.com/ghostty-org/ghostty/discussions/3224#discussioncomment-11684863) doesn't change that

**Reply by voidastro4** [reply]
enabling io_uring just for a terminal application is sadly not a solution.

**Reply by 00-kat** [reply]
@ssergiienko 

> even with **epoll** and **uring disabled**

Are you sure?  Check with `ghostty --version`.

I found that, for some reason, if I built Ghostty with the `ghostty-git` AUR package, the change wasn't applying for some reason, so I had to run it manually from another terminal window while Zig was compiling other packages for it to work.  Maybe try that, instead of running it before compilation?

And if it indeed is using Epoll, strace ghostty to see which syscall fails with EPERM.  In the case of `io_uring`, the `io_uring_setup` syscall would fail.

**Reply by ssergiienko** [reply]
@00-kat one more person confirmed the same https://github.com/ghostty-org/ghostty/discussions/3224#discussioncomment-11687922

**Reply by 00-kat** [reply]
Yeah I saw that.  @voidastro4 and @ssergiienko can you both please try what I mentioned in https://github.com/ghostty-org/ghostty/discussions/3267#discussioncomment-11687692?

**Reply by voidastro4** [reply]
> Yeah I saw that. @voidastro4 and @ssergiienko can you both please try what I mentioned in [#3267 (reply in thread)](https://github.com/ghostty-org/ghostty/discussions/3267#discussioncomment-11687692)?

Yes globbing was the problem. It works.

Hopefully the fallback will be present in future ghostty releases rather than crash without io_uring and have users compile their own versions.

**Reply by 00-kat** [reply]
Oh I see, I had thought bash enables recursive globbing by default.  I'll edit my original comment to add a note ^^.

**Reply by Mikilio** [reply]
Any chance this gets changed in the code? I doubt it's giving that much performance, and enough systems have io_uring disabled.

**Reply by 00-kat** [reply]
@Mikilio do you mean https://github.com/mitchellh/libxev/issues/131 or do you mean a build option to switch instead of having to manually patch the code?

**Reply by Mikilio** [reply]
I mean the code of ghostty. I was actually suggesting to go even further and use epolling per default. But a build option may be better.

**Reply by 00-kat** [reply]
Unless there's incentive to (e.g. if something is broken), a user is unlikely to find any features (such as `io_uring`) disabled by default: that is to say, making Epoll the default is almost equivalent to not having `io_uring` whatsoever.  I have heard that `io_uring` results in better performance than Epoll though, which is probably the reason why it is used.

TL;DR: in my opinion a build option is better.

**Comment by 00-kat** [comment]
Note: this is no longer required since #5916; it'll just work™ now.
